File: src/robot-localization(13)/src/odom_tf_fusion_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <geometry_msgs/msg/pose_with_covariance_stamped.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <tf2_msgs/msg/tf_message.hpp>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <Eigen/Dense>
#include <cmath>

class OdomTFFusionNode : public rclcpp::Node
{
public:
    OdomTFFusionNode() : Node("odom_tf_fusion_node")
    {
        // Initialize EKF state: [x, y, theta, vx, vy, omega]
        state_ = Eigen::VectorXd::Zero(6);
        P_ = Eigen::MatrixXd::Identity(6, 6) * 0.1;
        Q_ = Eigen::MatrixXd::Zero(6, 6);
        Q_.diagonal() << 0.01, 0.01, 0.01, 0.1, 0.1, 0.1;
        
        R_odom_ = Eigen::MatrixXd::Zero(3, 3);
        R_odom_.diagonal() << 0.02, 0.02, 0.01;
        
        R_tf_ = Eigen::MatrixXd::Zero(3, 3);
        R_tf_.diagonal() << 0.03, 0.03, 0.015;
        
        // QoS profiles
        auto sensor_qos = rclcpp::QoS(10).best_effort();
        auto reliable_qos = rclcpp::QoS(10).reliable();
        
        // Subscribers
        odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
            "/odom", sensor_qos,
            std::bind(&OdomTFFusionNode::odomCallback, this, std::placeholders::_1));
        
        tf_sub_ = this->create_subscription<tf2_msgs::msg::TFMessage>(
            "/tf", sensor_qos,
            std::bind(&OdomTFFusionNode::tfCallback, this, std::placeholders::_1));
        
        // Publishers
        fused_odom_pub_ = this->create_publisher<nav_msgs::msg::Odometry>(
            "/localization/odometry", reliable_qos);
        
        pose_pub_ = this->create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>(
            "/localization/pose", reliable_qos);
        
        // TF broadcaster
        tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);
        
        // Timer for main fusion loop (50 Hz)
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(20),
            std::bind(&OdomTFFusionNode::fusionCallback, this));
        
        last_update_time_ = this->now();
        
        // Fusion weights
        odom_weight_ = 0.7;
        tf_weight_ = 0.3;
        
        RCLCPP_INFO(this->get_logger(), "Odometry-TF Fusion Node initialized");
        RCLCPP_INFO(this->get_logger(), "Subscribing to: /odom, /tf");
        RCLCPP_INFO(this->get_logger(), "Fusion weights: odom=%.1f, tf=%.1f", odom_weight_, tf_weight_);
    }

private:
    void odomCallback(const nav_msgs::msg::Odometry::SharedPtr msg)
    {
        latest_odom_ = msg;
        RCLCPP_DEBUG_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
            "Odom: pos=(%.3f, %.3f), vel=(%.3f, %.3f)",
            msg->pose.pose.position.x, msg->pose.pose.position.y,
            msg->twist.twist.linear.x, msg->twist.twist.angular.z);
    }
    
    void tfCallback(const tf2_msgs::msg::TFMessage::SharedPtr msg)
    {
        for (const auto& transform : msg->transforms) {
            if (transform.header.frame_id == "Assem1/odom" && 
                transform.child_frame_id == "Assem1/base_link") {
                
                latest_tf_transform_ = std::make_shared<geometry_msgs::msg::TransformStamped>(transform);
                last_tf_time_ = this->now();
                
                RCLCPP_DEBUG_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
                    "TF: pos=(%.3f, %.3f), frame=%s -> %s",
                    transform.transform.translation.x, transform.transform.translation.y,
                    transform.header.frame_id.c_str(), transform.child_frame_id.c_str());
                break;
            }
        }
    }
    
    void predict(double dt)
    {
        if (dt <= 0) return;
        
        Eigen::MatrixXd F = Eigen::MatrixXd::Identity(6, 6);
        F(0, 3) = dt;
        F(1, 4) = dt;
        F(2, 5) = dt;
        
        state_ = F * state_;
        state_(2) = normalizeAngle(state_(2));
        P_ = F * P_ * F.transpose() + Q_;
    }
    
    void fusionCallback()
    {
        if (!latest_odom_) return;
        
        auto current_time = this->now();
        double dt = (current_time - last_update_time_).seconds();
        
        if (dt > 0) {
            predict(dt);
            
            // Extract odometry data
            double odom_x = latest_odom_->pose.pose.position.x;
            double odom_y = latest_odom_->pose.pose.position.y;
            
            tf2::Quaternion odom_q(
                latest_odom_->pose.pose.orientation.x,
                latest_odom_->pose.pose.orientation.y,
                latest_odom_->pose.pose.orientation.z,
                latest_odom_->pose.pose.orientation.w);
            
            double roll, pitch, yaw;
            tf2::Matrix3x3(odom_q).getRPY(roll, pitch, yaw);
            double odom_theta = yaw;
            
            double odom_vx = latest_odom_->twist.twist.linear.x;
            double odom_vy = latest_odom_->twist.twist.linear.y;
            double odom_omega = latest_odom_->twist.twist.angular.z;
            
            // Update with odometry
            updateOdometry(odom_x, odom_y, odom_theta, odom_vx, odom_vy, odom_omega);
            
            // Update with TF if available and recent
            if (latest_tf_transform_ && last_tf_time_.has_value()) {
                double tf_age = (current_time - last_tf_time_.value()).seconds();
                if (tf_age < 0.5) {  // Only use TF data if it's less than 500ms old
                    updateWithTF();
                }
            }
            
            publishResults();
            last_update_time_ = current_time;
        }
    }
    
    void updateOdometry(double x, double y, double theta, double vx, double vy, double omega)
    {
        Eigen::MatrixXd H = Eigen::MatrixXd::Identity(6, 6);
        
        Eigen::VectorXd z(6);
        z << x, y, theta, vx, vy, omega;
        
        Eigen::VectorXd y_k = z - H * state_;
        y_k(2) = normalizeAngle(y_k(2));
        
        Eigen::MatrixXd R_full = Eigen::MatrixXd::Zero(6, 6);
        R_full.block<3, 3>(0, 0) = R_odom_;
        R_full.block<3, 3>(3, 3) = Eigen::Vector3d(0.05, 0.05, 0.02).asDiagonal();
        
        Eigen::MatrixXd S = H * P_ * H.transpose() + R_full;
        Eigen::MatrixXd K = P_ * H.transpose() * S.inverse();
        
        state_ = state_ + K * y_k;
        state_(2) = normalizeAngle(state_(2));
        P_ = (Eigen::MatrixXd::Identity(6, 6) - K * H) * P_;
    }
    
    void updateWithTF()
    {
        if (!latest_tf_transform_) return;
        
        double tf_x = latest_tf_transform_->transform.translation.x;
        double tf_y = latest_tf_transform_->transform.translation.y;
        
        tf2::Quaternion tf_q(
            latest_tf_transform_->transform.rotation.x,
            latest_tf_transform_->transform.rotation.y,
            latest_tf_transform_->transform.rotation.z,
            latest_tf_transform_->transform.rotation.w);
        
        double tf_roll, tf_pitch, tf_yaw;
        tf2::Matrix3x3(tf_q).getRPY(tf_roll, tf_pitch, tf_yaw);
        
        Eigen::MatrixXd H_tf = Eigen::MatrixXd::Zero(3, 6);
        H_tf(0, 0) = 1;
        H_tf(1, 1) = 1;
        H_tf(2, 2) = 1;
        
        Eigen::VectorXd z_tf(3);
        z_tf << tf_x, tf_y, tf_yaw;
        
        Eigen::VectorXd y_k_tf = z_tf - H_tf * state_;
        y_k_tf(2) = normalizeAngle(y_k_tf(2));
        
        Eigen::MatrixXd S_tf = H_tf * P_ * H_tf.transpose() + R_tf_;
        Eigen::MatrixXd K_tf = P_ * H_tf.transpose() * S_tf.inverse();
        
        state_ = state_ + K_tf * y_k_tf;
        state_(2) = normalizeAngle(state_(2));
        P_ = (Eigen::MatrixXd::Identity(6, 6) - K_tf * H_tf) * P_;
    }
    
    void publishResults()
    {
        publishFusedOdometry();
        publishPose();
        publishTransform();
    }
    
    void publishFusedOdometry()
    {
        auto msg = nav_msgs::msg::Odometry();
        msg.header.stamp = this->now();
        msg.header.frame_id = "odom";
        msg.child_frame_id = "base_link";
        
        msg.pose.pose.position.x = state_(0);
        msg.pose.pose.position.y = state_(1);
        msg.pose.pose.position.z = 0.0;
        
        tf2::Quaternion q;
        q.setRPY(0, 0, state_(2));
        msg.pose.pose.orientation = tf2::toMsg(q);
        
        msg.twist.twist.linear.x = state_(3);
        msg.twist.twist.linear.y = state_(4);
        msg.twist.twist.angular.z = state_(5);
        
        fused_odom_pub_->publish(msg);
    }
    
    void publishPose()
    {
        auto msg = geometry_msgs::msg::PoseWithCovarianceStamped();
        msg.header.stamp = this->now();
        msg.header.frame_id = "odom";
        
        msg.pose.pose.position.x = state_(0);
        msg.pose.pose.position.y = state_(1);
        msg.pose.pose.position.z = 0.0;
        
        tf2::Quaternion q;
        q.setRPY(0, 0, state_(2));
        msg.pose.pose.orientation = tf2::toMsg(q);
        
        std::fill(msg.pose.covariance.begin(), msg.pose.covariance.end(), 0.0);
        msg.pose.covariance[0] = P_(0, 0);
        msg.pose.covariance[7] = P_(1, 1);
        msg.pose.covariance[35] = P_(2, 2);
        
        pose_pub_->publish(msg);
    }
    
    void publishTransform()
    {
        geometry_msgs::msg::TransformStamped t;
        t.header.stamp = this->now();
        t.header.frame_id = "odom";
        t.child_frame_id = "base_link";
        
        t.transform.translation.x = state_(0);
        t.transform.translation.y = state_(1);
        t.transform.translation.z = 0.0;
        
        tf2::Quaternion q;
        q.setRPY(0, 0, state_(2));
        t.transform.rotation = tf2::toMsg(q);
        
        tf_broadcaster_->sendTransform(t);
    }
    
    double normalizeAngle(double angle)
    {
        return atan2(sin(angle), cos(angle));
    }
    
    // Member variables
    Eigen::VectorXd state_;
    Eigen::MatrixXd P_, Q_, R_odom_, R_tf_;
    
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
    rclcpp::Subscription<tf2_msgs::msg::TFMessage>::SharedPtr tf_sub_;
    rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr fused_odom_pub_;
    rclcpp::Publisher<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr pose_pub_;
    
    std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
    rclcpp::TimerBase::SharedPtr timer_;
    
    nav_msgs::msg::Odometry::SharedPtr latest_odom_;
    std::shared_ptr<geometry_msgs::msg::TransformStamped> latest_tf_transform_;
    rclcpp::Time last_update_time_;
    std::optional<rclcpp::Time> last_tf_time_;
    
    double odom_weight_, tf_weight_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<OdomTFFusionNode>());
    rclcpp::shutdown();
    return 0;
}

File: src/robot-localization(13)/robot_localization/odom_tf_fusion_node.py
#!/usr/bin/env python3

import numpy as np
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy
from geometry_msgs.msg import PoseWithCovarianceStamped, TransformStamped
from nav_msgs.msg import Odometry
from tf2_msgs.msg import TFMessage
from tf2_ros import TransformBroadcaster
import tf_transformations
import math

class OdomTFFusionNode(Node):
    """
    Localization node that fuses:
    - /odom (already fused dual encoder data)
    - /tf (transform data)
    
    Uses Extended Kalman Filter for optimal sensor fusion
    """
    
    def __init__(self):
        super().__init__('odom_tf_fusion_node')
        
        # EKF state: [x, y, theta, vx, vy, omega]
        self.state = np.zeros(6)
        self.P = np.eye(6) * 0.1  # Covariance matrix
        self.Q = np.diag([0.01, 0.01, 0.01, 0.1, 0.1, 0.1])  # Process noise
        self.R_odom = np.diag([0.02, 0.02, 0.01])  # Odometry noise (lower since it's already fused)
        self.R_tf = np.diag([0.03, 0.03, 0.015])  # TF noise (slightly higher uncertainty)
        
        # QoS profiles
        self.sensor_qos = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            history=HistoryPolicy.KEEP_LAST,
            depth=10
        )
        
        self.reliable_qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            history=HistoryPolicy.KEEP_LAST,
            depth=10
        )
        
        # Subscribers
        self.odom_sub = self.create_subscription(
            Odometry,
            '/odom',  # Your dual-encoder fused odometry
            self.odom_callback,
            self.sensor_qos
        )
        
        self.tf_sub = self.create_subscription(
            TFMessage,
            '/tf',  # Your transform data
            self.tf_callback,
            self.sensor_qos
        )
        
        # Publishers
        self.fused_odom_pub = self.create_publisher(
            Odometry,
            '/localization/odometry',
            self.reliable_qos
        )
        
        self.pose_pub = self.create_publisher(
            PoseWithCovarianceStamped,
            '/localization/pose',
            self.reliable_qos
        )
        
        # TF broadcaster for output
        self.tf_broadcaster = TransformBroadcaster(self)
        
        # Data storage
        self.latest_odom = None
        self.latest_tf_transform = None
        self.last_update_time = self.get_clock().now()
        self.last_tf_time = None
        
        # Fusion parameters (adjustable)
        self.odom_weight = 0.7  # Higher weight for encoder data
        self.tf_weight = 0.3    # Lower weight for TF data
        
        # Timer for main fusion loop
        self.timer = self.create_timer(0.02, self.fusion_callback)  # 50 Hz
        
        self.get_logger().info('Odometry-TF Fusion Node initialized')
        self.get_logger().info('Subscribing to:')
        self.get_logger().info('  - /odom (nav_msgs/Odometry) - dual encoder fused data')
        self.get_logger().info('  - /tf (tf2_msgs/TFMessage) - transform data')
        self.get_logger().info('Publishing to:')
        self.get_logger().info('  - /localization/odometry (nav_msgs/Odometry)')
        self.get_logger().info('  - /localization/pose (geometry_msgs/PoseWithCovarianceStamped)')
        self.get_logger().info(f'Fusion weights: odom={self.odom_weight}, tf={self.tf_weight}')
    
    def odom_callback(self, msg):
        """Process odometry from dual encoder fusion"""
        self.latest_odom = msg
        
        self.get_logger().debug(
            f'Odom: pos=({msg.pose.pose.position.x:.3f}, {msg.pose.pose.position.y:.3f}), '
            f'vel=({msg.twist.twist.linear.x:.3f}, {msg.twist.twist.angular.z:.3f})',
            throttle_duration_sec=2.0
        )
    
    def tf_callback(self, msg):
        """Process TF messages - look for your specific transform"""
        # Look for transforms from Assem1/odom to Assem1/base_link (from your data)
        for transform in msg.transforms:
            if (transform.header.frame_id == 'Assem1/odom' and 
                transform.child_frame_id == 'Assem1/base_link'):
                
                self.latest_tf_transform = transform
                self.last_tf_time = self.get_clock().now()
                
                self.get_logger().debug(
                    f'TF: pos=({transform.transform.translation.x:.3f}, '
                    f'{transform.transform.translation.y:.3f}), '
                    f'frame={transform.header.frame_id} -> {transform.child_frame_id}',
                    throttle_duration_sec=2.0
                )
                break
    
    def predict(self, dt):
        """EKF prediction step"""
        if dt <= 0:
            return
            
        # State transition matrix
        F = np.eye(6)
        F[0, 3] = dt  # x += vx * dt
        F[1, 4] = dt  # y += vy * dt  
        F[2, 5] = dt  # theta += omega * dt
        
        # Predict state
        self.state = F @ self.state
        self.state[2] = self.normalize_angle(self.state[2])
        
        # Predict covariance
        self.P = F @ self.P @ F.T + self.Q
    
    def fusion_callback(self):
        """Main fusion callback - combines odometry and TF data"""
        if self.latest_odom is None:
            return
        
        current_time = self.get_clock().now()
        dt = (current_time - self.last_update_time).nanoseconds / 1e9
        
        if dt > 0:
            # Prediction step
            self.predict(dt)
            
            # Extract odometry data
            odom_x = self.latest_odom.pose.pose.position.x
            odom_y = self.latest_odom.pose.pose.position.y
            odom_q = self.latest_odom.pose.pose.orientation
            odom_euler = tf_transformations.euler_from_quaternion([odom_q.x, odom_q.y, odom_q.z, odom_q.w])
            odom_theta = odom_euler[2]
            
            # Extract velocity data
            odom_vx = self.latest_odom.twist.twist.linear.x
            odom_vy = self.latest_odom.twist.twist.linear.y
            odom_omega = self.latest_odom.twist.twist.angular.z
            
            # Update with odometry measurement
            self.update_odometry(odom_x, odom_y, odom_theta, odom_vx, odom_vy, odom_omega)
            
            # Update with TF data if available and recent
            if self.latest_tf_transform is not None and self.last_tf_time is not None:
                tf_age = (current_time - self.last_tf_time).nanoseconds / 1e9
                if tf_age < 0.5:  # Only use TF data if it's less than 500ms old
                    self.update_with_tf()
                else:
                    self.get_logger().debug('TF data too old, skipping TF update', throttle_duration_sec=5.0)
            
            # Publish fused results
            self.publish_fused_odometry()
            self.publish_pose()
            self.publish_transform()
            
            self.last_update_time = current_time
    
    def update_odometry(self, x, y, theta, vx, vy, omega):
        """Update EKF with odometry measurement"""
        # Measurement model for position, orientation, and velocities
        H = np.eye(6)  # Direct observation of all states
        
        z = np.array([x, y, theta, vx, vy, omega])
        y_k = z - H @ self.state
        y_k[2] = self.normalize_angle(y_k[2])  # Handle angle wrapping
        
        # Measurement noise covariance
        R_full = np.zeros((6, 6))
        R_full[:3, :3] = self.R_odom  # Position and orientation
        R_full[3:, 3:] = np.diag([0.05, 0.05, 0.02])  # Velocity measurements
        
        S = H @ self.P @ H.T + R_full
        K = self.P @ H.T @ np.linalg.inv(S)
        
        self.state = self.state + K @ y_k
        self.state[2] = self.normalize_angle(self.state[2])
        self.P = (np.eye(6) - K @ H) @ self.P
    
    def update_with_tf(self):
        """Update EKF with TF transform data"""
        if self.latest_tf_transform is None:
            return
        
        # Extract TF data
        tf_x = self.latest_tf_transform.transform.translation.x
        tf_y = self.latest_tf_transform.transform.translation.y
        tf_q = self.latest_tf_transform.transform.rotation
        tf_euler = tf_transformations.euler_from_quaternion([tf_q.x, tf_q.y, tf_q.z, tf_q.w])
        tf_theta = tf_euler[2]
        
        # Measurement model for TF (position and orientation only)
        H_tf = np.zeros((3, 6))
        H_tf[0, 0] = 1  # x
        H_tf[1, 1] = 1  # y
        H_tf[2, 2] = 1  # theta
        
        z_tf = np.array([tf_x, tf_y, tf_theta])
        y_k_tf = z_tf - H_tf @ self.state
        y_k_tf[2] = self.normalize_angle(y_k_tf[2])  # Handle angle wrapping
        
        S_tf = H_tf @ self.P @ H_tf.T + self.R_tf
        K_tf = self.P @ H_tf.T @ np.linalg.inv(S_tf)
        
        self.state = self.state + K_tf @ y_k_tf
        self.state[2] = self.normalize_angle(self.state[2])
        self.P = (np.eye(6) - K_tf @ H_tf) @ self.P
        
        self.get_logger().debug(
            f'TF Update: tf_pos=({tf_x:.3f}, {tf_y:.3f}), '
            f'fused_pos=({self.state[0]:.3f}, {self.state[1]:.3f})',
            throttle_duration_sec=2.0
        )
    
    def publish_fused_odometry(self):
        """Publish fused odometry result"""
        msg = Odometry()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = 'odom'
        msg.child_frame_id = 'base_link'
        
        # Position from EKF state
        msg.pose.pose.position.x = self.state[0]
        msg.pose.pose.position.y = self.state[1]
        msg.pose.pose.position.z = 0.0
        
        # Orientation from EKF state
        quat = tf_transformations.quaternion_from_euler(0, 0, self.state[2])
        msg.pose.pose.orientation.x = quat[0]
        msg.pose.pose.orientation.y = quat[1]
        msg.pose.pose.orientation.z = quat[2]
        msg.pose.pose.orientation.w = quat[3]
        
        # Velocity from EKF state
        msg.twist.twist.linear.x = self.state[3]
        msg.twist.twist.linear.y = self.state[4]
        msg.twist.twist.angular.z = self.state[5]
        
        # Covariance from EKF
        pose_cov = np.zeros(36)
        pose_cov[0] = self.P[0, 0]   # x variance
        pose_cov[7] = self.P[1, 1]   # y variance
        pose_cov[35] = self.P[2, 2]  # theta variance
        msg.pose.covariance = pose_cov.tolist()
        
        twist_cov = np.zeros(36)
        twist_cov[0] = self.P[3, 3]   # vx variance
        twist_cov[7] = self.P[4, 4]   # vy variance
        twist_cov[35] = self.P[5, 5]  # omega variance
        msg.twist.covariance = twist_cov.tolist()
        
        self.fused_odom_pub.publish(msg)
    
    def publish_pose(self):
        """Publish pose with covariance"""
        msg = PoseWithCovarianceStamped()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = 'odom'
        
        # Position
        msg.pose.pose.position.x = self.state[0]
        msg.pose.pose.position.y = self.state[1]
        msg.pose.pose.position.z = 0.0
        
        # Orientation
        quat = tf_transformations.quaternion_from_euler(0, 0, self.state[2])
        msg.pose.pose.orientation.x = quat[0]
        msg.pose.pose.orientation.y = quat[1]
        msg.pose.pose.orientation.z = quat[2]
        msg.pose.pose.orientation.w = quat[3]
        
        # Covariance
        pose_cov = np.zeros(36)
        pose_cov[0] = self.P[0, 0]
        pose_cov[7] = self.P[1, 1]
        pose_cov[35] = self.P[2, 2]
        msg.pose.covariance = pose_cov.tolist()
        
        self.pose_pub.publish(msg)
    
    def publish_transform(self):
        """Publish fused transform"""
        t = TransformStamped()
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = 'odom'
        t.child_frame_id = 'base_link'
        
        # Translation
        t.transform.translation.x = self.state[0]
        t.transform.translation.y = self.state[1]
        t.transform.translation.z = 0.0
        
        # Rotation
        quat = tf_transformations.quaternion_from_euler(0, 0, self.state[2])
        t.transform.rotation.x = quat[0]
        t.transform.rotation.y = quat[1]
        t.transform.rotation.z = quat[2]
        t.transform.rotation.w = quat[3]
        
        self.tf_broadcaster.sendTransform(t)
    
    def normalize_angle(self, angle):
        """Normalize angle to [-pi, pi]"""
        return math.atan2(math.sin(angle), math.cos(angle))

def main(args=None):
    rclpy.init(args=args)
    
    try:
        node = OdomTFFusionNode()
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == '__main__':
    main()
